--- a/pqueue.h
+++ b/pqueue.h
@@ -5,11 +5,11 @@
 #include <sys/time.h>
 
 /* wait this many seconds for missing packets before forgetting about them */
-#define DEFAULT_PACKET_TIMEOUT 0.3
+#define DEFAULT_PACKET_TIMEOUT 0.4
 extern int packet_timeout_usecs;
 
 /* assume packet is bad/spoofed if it's more than this many seqs ahead */
-#define MISSING_WINDOW 300
+#define MISSING_WINDOW 500
 
 /* Packet queue structure: linked list of packets received out-of-order */
 typedef struct pqueue {
